public/app/features/explore/RichHistory/RichHistoryQueriesTab.tsx

```diff
@@ -132,50 +137,49 @@ export function RichHistoryQueriesTab(props: Props) {
     height,
   } = props;
 
-  const [sliderRetentionFilter, setSliderRetentionFilter] = useState<[number, number]>([0, retentionPeriod]);
+  const [timeFilter, setTimeFilter] = useState<[number, number]>([0, retentionPeriod]);
+  const [filteredQueries, setFilteredQueries] = useState([]);
+  const [searchInput, setSearchInput] = useState('');
 
   const theme = useTheme();
   const styles = getStyles(theme, height);
+
   const datasourcesRetrievedFromQueryHistory = uniqBy(queries, 'datasourceName').map(d => d.datasourceName);
   const listOfDatasources = createDatasourcesList(datasourcesRetrievedFromQueryHistory);
-
   const listOfDatasourceFilters = datasourceFilters?.map(d => d.value);
-  const filteredQueriesByDatasource =
-    listOfDatasourceFilters && listOfDatasourceFilters?.length > 0
-      ? queries?.filter(q => listOfDatasourceFilters?.includes(q.datasourceName))
-      : queries;
-
-  const sortedQueries = sortQueries(filteredQueriesByDatasource, sortOrder);
-  const queriesWithinSelectedTimeline = sortedQueries?.filter(
-    q =>
-      q.ts < createRetentionPeriodBoundary(sliderRetentionFilter[0], true) &&
-      q.ts > createRetentionPeriodBoundary(sliderRetentionFilter[1], false)
-  );
+
+  useEffect(() => {
+    setFilteredQueries(filterAndSortQueries(queries, sortOrder, listOfDatasourceFilters, searchInput, timeFilter));
+  }, [queries, sortOrder, datasourceFilters]);
+
+  const filterAndSortQueriesDebounced = debounce(searchValue => {
+    setFilteredQueries(filterAndSortQueries(queries, sortOrder, listOfDatasourceFilters, searchValue, timeFilter));
+  }, 300);
```

Comment:
This actually doesn't seem to work (you can try by mashing your keys - it obviously shouldn't update) because the debounced function is recreated on every render. I would recommend using `use-debounce`, it makes debouncing handlers with hooks so easy...
Unless of course you find common debouncing patterns elsewhere in the code base that are just as straight-forward.

Reply:
Yes you are right. 💙 Thanks so much for suggesting `use-debounce`. Fixed, ready for review.

Reply:
Debouncing looks good now! :)

Reply:
⛹️‍♀️⛹️‍♀️⛹️‍♀️